src/module.c

```diff
@@ -1109,213 +1080,667 @@ int RM_CreateSubcommand(RedisModuleCommand *parent, const char *name, RedisModul
-        return REDISMODULE_ERR;
+    if (info->tips) {
+        size_t count = 0;
+        for (count = 0; info->tips[count]; count++);
+        serverAssert(count < SIZE_MAX / sizeof(char *));
+        cmd->tips = zmalloc(sizeof(char *) * (count + 1));
+        for (size_t j = 0; j < count; j++) cmd->tips[j] = zstrdup(info->tips[j]);
+        cmd->tips[count] = NULL;
+        cmd->num_tips = count;
+    }
 
-    struct redisCommand *cmd = command->rediscmd;
+    if (info->arity) cmd->arity = info->arity;
 
-    extendKeySpecsIfNeeded(cmd);
+    if (info->key_specs) {
+        /* Count and allocate the key specs. */
+        size_t count;
+        for (count = 0; info->key_specs[count].begin_search_type; count++);
+        serverAssert(count < INT_MAX);
+        if (count <= STATIC_KEY_SPECS_NUM) {
+            cmd->key_specs_max = STATIC_KEY_SPECS_NUM;
+            cmd->key_specs = cmd->key_specs_static;
+        } else {
+            cmd->key_specs_max = count;
+            cmd->key_specs = zmalloc(sizeof(keySpec) * count);
+        }
 
-    *index = cmd->key_specs_num;
-    cmd->key_specs[cmd->key_specs_num].begin_search_type = KSPEC_BS_INVALID;
-    cmd->key_specs[cmd->key_specs_num].find_keys_type = KSPEC_FK_INVALID;
-    cmd->key_specs[cmd->key_specs_num].flags = flags;
-    cmd->key_specs_num++;
-    return REDISMODULE_OK;
-}
+        /* Copy the contents of the RedisModuleCommandKeySpec array. */
+        cmd->key_specs_num = count;
+        for (size_t j = 0; j < count; j++) {
+            cmd->key_specs[j].flags = moduleConvertKeySpecsFlags(info->key_specs[j].flags);
+            switch (info->key_specs[j].begin_search_type) {
+            case REDISMODULE_KSPEC_BS_UNKNOWN:
+                cmd->key_specs[j].begin_search_type = KSPEC_BS_UNKNOWN;
+                break;
+            case REDISMODULE_KSPEC_BS_INDEX:
+                cmd->key_specs[j].begin_search_type = KSPEC_BS_INDEX;
+                cmd->key_specs[j].bs.index.pos = info->key_specs[j].bs.index.pos;
+                break;
+            case REDISMODULE_KSPEC_BS_KEYWORD:
+                cmd->key_specs[j].begin_search_type = KSPEC_BS_KEYWORD;
+                cmd->key_specs[j].bs.keyword.keyword = zstrdup(info->key_specs[j].bs.keyword.keyword);
+                cmd->key_specs[j].bs.keyword.startfrom = info->key_specs[j].bs.keyword.startfrom;
+                break;
+            default:
+                serverAssert(0);
+            }
 
-int moduleSetCommandKeySpecBeginSearch(RedisModuleCommand *command, int index, keySpec *spec) {
-    struct redisCommand *cmd = command->rediscmd;
+            switch (info->key_specs[j].find_keys_type) {
+            case 0:
+                /* Omitted. This is shorthand to say that it's a single key. */
+                cmd->key_specs[j].find_keys_type = KSPEC_FK_RANGE;
+                cmd->key_specs[j].fk.range.lastkey = 0;
+                cmd->key_specs[j].fk.range.keystep = 1;
+                cmd->key_specs[j].fk.range.limit = 0;
```

Comment:
ohh, now i understand `REDISMODULE_KSPEC_FK_OMITTED).
would have been better to use it here. (and in the other `case 0:`).
and comment in the declaration of `FK_OMITTED` that it has to be 0 for that reason.

i now realize that you want the module to rely on the struct initialization to 0, rather than explicitly use `FK_OMITTED` in it's code

i suppose that for many modules that need just one key in a constant position, that shorthand is saving a few chars.
but also, for these, there's actually no need to define the key spec, right?
the legacy keyspec set by RM_CreateCommand, so i'm not sure there's a lot of value in that trick.

Reply:
Keyspecs are only needed to set flags in that case.  Don't we want modules to provide flags?

Reply:
yes, we do... we must: https://github.com/redis/redis/issues/10144
i suppose i'm ok with this OMITTED thing. just needs some additional comments.